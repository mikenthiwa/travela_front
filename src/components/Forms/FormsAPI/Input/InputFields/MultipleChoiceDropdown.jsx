import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './MultiChoiceDropdown.scss';
import selectDropdownIcon from '../../../../../images/icons/form_select_dropdown.svg';

export default class MultipleChoiceDropdown extends Component {
  state = {
    open: false,
    selectedOption: 0,
    value: ''
  };
  dropdown = React.createRef();
  input = React.createRef();

  handleMenuClick = (index) => {
    this.setState({ selectedOption: index});
  };

  componentDidMount = () => {
    document.addEventListener('click', this.hideDropdownOnClickOutside);
  };

  hideDropdownOnClickOutside = (e) => {
    if(this.dropdown && e.target && !this.dropdown.current.contains(e.target)){
      this.setState({ open: false});
    }
  };

  componentWillUnmount = () => {
    document.removeEventListener('click', this.hideDropdownOnClickOutside);
  };

  handleChoice = (value) => {
    const { onChange, options } = this.props;
    this.setState({ value, open: false }, () => {
      const { selectedOption, value } = this.state;
      onChange({ target: {
        ...this.input.current,
        value: {
          option: options[selectedOption].name,
          value
        }
      }});
    });
  };

  handleClick = () => {
    this.setState((state) => {
      return {
        open: !state.open
      };
    });
  };

  renderDropdown() {
    const { selectedOption, value, open } = this.state;
    const { options } = this.props;

    return (
      <div className={`multichoice-dropdown__dropdown ${open ? 'open': 'closed'}`}>
        <ul className="multichoice-dropdown__dropdown__menu">
          {options.map((option, index) => (
            <li
              role="presentation"
              className={`${index === selectedOption ? 'active' : ''}`}
              key={option.name}
              onClick={() => this.handleMenuClick(index)}
            >
              {option.name}
            </li>
          ))}
        </ul>
        <ul className="multichoice-dropdown__dropdown__list">
          {options[selectedOption].items &&
            options[selectedOption].items.map((item) => (
              <li
                key={item + Math.random()}
                className={`${item === value ? 'active': ''}`}
                role="presentation"
                onClick={() => this.handleChoice(item)}>
                {item}
              </li>
            ))}
        </ul>
      </div>
    );
  }

  render() {
    const { open, value } = this.state;
    const { onChange } = this.props;
    return (
      <div
        ref={this.dropdown}
        className={`multichoice-dropdown ${open ? 'open': 'closed'}`}
      >
        <div className="style-dropdown">
          <input
            type="text" value={value}
            ref={this.input}
            id="flight-input"
            onChange={onChange}
            onClick={this.handleClick}
            autoComplete="off"
          />
          <img
            src={selectDropdownIcon} alt="icn"
            onClick={this.handleClick}
            role="presentation"
          />
        </div>
        {this.renderDropdown()}
      </div>
    );
  }
}
MultipleChoiceDropdown.propTypes = {
  onChange: PropTypes.func,
  options: PropTypes.array.isRequired
};
MultipleChoiceDropdown.defaultProps = {
  onChange: () => {}
};
