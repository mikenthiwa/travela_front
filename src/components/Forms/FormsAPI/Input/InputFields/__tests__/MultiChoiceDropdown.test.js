import React from 'react';
import { shallow, mount } from 'enzyme';
import MultipleChoiceDropdown from '../MultipleChoiceDropdown';

describe('<MultipleChoiceDropdown />', () => {
  const props = {
    options: [
      {
        name: 'Countries',
        items: ["Aruba"]
      }
    ],
    selectedOption: 0,
    value: 'France',
    handleMenuClick: jest.fn(),
    handleClick: jest.fn(),
    handleChoice: jest.fn(),
    hideDropdownOnClickOutside: jest.fn(),
    onChange: jest.fn()
  };

  it('should render without crashing', () => {
    const wrapper = shallow(<MultipleChoiceDropdown {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle click', () => {
    const wrapper = mount(<MultipleChoiceDropdown {...props} />);
    const instance = wrapper.instance();
    wrapper.setState({
      open: false
    });
    const menu = wrapper.find('input');
    menu.simulate('click');
    expect(instance.state.open).toBe(true);
  });

  it('should handle menu click', () => {
    const wrapper = mount(<MultipleChoiceDropdown {...props} />);
    const instance = wrapper.instance();
    wrapper.setState({
      selectedOption: 0
    });
    const menu = wrapper.find('li').first();
    menu.simulate('click');
    expect(instance.state.selectedOption).toBe(0);
  });

  it('should handle hiding the dropdown onclicking outside', () => {
    const wrapper = mount(<MultipleChoiceDropdown {...props} />);
    const e = {target:{
      name:""
    }}
    const div = wrapper.find('div').first();
    div.simulate('click', e );
    const instance = wrapper.instance();
    wrapper.setState({
      open: false
    });
    instance.hideDropdownOnClickOutside({ preventDefault: jest.fn() });
    instance.componentWillUnmount()
    expect(instance.state.open).toBe(false);
  });

  it('should handle choice', () => {
    const wrapper = mount(<MultipleChoiceDropdown {...props} />);
    const instance = wrapper.instance();
    wrapper.setState({
      open: false,
      value: {
        value: 'France',
        option: 'Countries'
      },
      options: [
        {
          items: [],
          name: 'Countries'
        }
      ]
    });
    const choice = wrapper.find('li').last();
    choice.simulate('click');
    instance.handleChoice(props.value);
    expect(instance.state.value).toBe('France');
  });

  it('should have onChange default props', () => {
    const result = MultipleChoiceDropdown.defaultProps.onChange();
    expect(result).toBe(undefined);
  });
});
