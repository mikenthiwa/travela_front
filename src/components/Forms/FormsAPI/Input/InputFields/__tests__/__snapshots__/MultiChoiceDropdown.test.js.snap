// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultipleChoiceDropdown /> should render without crashing 1`] = `
<div
  className="multichoice-dropdown closed"
>
  <div
    className="style-dropdown"
  >
    <input
      autoComplete="off"
      id="flight-input"
      onChange={[Function]}
      onClick={[Function]}
      type="text"
      value=""
    />
    <img
      alt="icn"
      onClick={[Function]}
      role="presentation"
      src="form_select_dropdown.svg"
    />
  </div>
  <div
    className="multichoice-dropdown__dropdown closed"
  >
    <ul
      className="multichoice-dropdown__dropdown__menu"
    >
      <li
        className="active"
        key="Countries"
        onClick={[Function]}
        role="presentation"
      >
        Countries
      </li>
    </ul>
    <ul
      className="multichoice-dropdown__dropdown__list"
    >
      <li
        className=""
        key="Aruba0.9097494202270933"
        onClick={[Function]}
        role="presentation"
      >
        Aruba
      </li>
    </ul>
  </div>
</div>
`;
